daemon/api_systems.go

```diff
@@ -272,3 +328,212 @@ func postSystemActionInstall(c *Command, systemLabel string, req *systemActionRe
 		return BadRequest("unsupported install step %q", req.Step)
 	}
 }
+
+func parseValidationSets(validationSets []string) ([]*asserts.AtSequence, error) {
+	sets := make([]*asserts.AtSequence, 0, len(validationSets))
+	for _, vs := range validationSets {
+		account, name, seq, err := snapasserts.ParseValidationSet(vs)
+		if err != nil {
+			return nil, err
+		}
+
+		assertion := asserts.AtSequence{
+			Type:        asserts.ValidationSetType,
+			SequenceKey: []string{release.Series, account, name},
+			Pinned:      seq > 0,
+			Sequence:    seq,
+			Revision:    asserts.RevisionNotKnown,
+		}
+
+		sets = append(sets, &assertion)
+	}
+
+	return sets, nil
+}
+
+func readValueFromForm(form *Form, key string) (string, *apiError) {
+	values := form.Values[key]
+	if len(values) != 1 {
+		return "", BadRequest("expected exactly one %q value in form", key)
+	}
+	return values[0], nil
+}
+
+func readValueFromFormWithDefault(form *Form, key string, defaultValue string) (string, *apiError) {
+	values := form.Values[key]
+	switch len(values) {
+	case 0:
+		return defaultValue, nil
+	case 1:
+		return values[0], nil
+	default:
+		return "", BadRequest("expected at most one %q value in form", key)
+	}
+}
+
+func postSystemActionCreateOffline(c *Command, form *Form) Response {
+	label, errRsp := readValueFromForm(form, "label")
+	if errRsp != nil {
+		return errRsp
+	}
+
+	testSystemStr, errRsp := readValueFromFormWithDefault(form, "test-system", "false")
+	if errRsp != nil {
+		return errRsp
+	}
+
+	testSystem, err := strconv.ParseBool(testSystemStr)
+	if err != nil {
+		return BadRequest(`cannot parse "test-system" value as boolean: %s`, testSystemStr)
+	}
+
+	markDefaultStr, errRsp := readValueFromFormWithDefault(form, "mark-default", "false")
+	if errRsp != nil {
+		return errRsp
+	}
+
+	markDefault, err := strconv.ParseBool(markDefaultStr)
+	if err != nil {
+		return BadRequest(`cannot parse "mark-default" value as boolean: %s`, markDefaultStr)
+	}
+
+	vsetsList, errRsp := readValueFromFormWithDefault(form, "validation-sets", "")
+	if errRsp != nil {
+		return errRsp
+	}
+
+	var splitVSets []string
+	if vsetsList != "" {
+		splitVSets = strings.Split(vsetsList, ",")
+	}
+
+	// TODO: why is this a comma delimited list, rather than multiple values?
+	sequences, err := parseValidationSets(splitVSets)
+	if err != nil {
+		return BadRequest("cannot parse validation sets: %v", err)
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var snapFiles []*uploadedSnap
+	if len(form.FileRefs["snap"]) > 0 {
```

Comment:
Could we read the values from form before we actually take the state lock, specifically `"snap" and "assertion"` form members